pkg/crds/crds.go

```diff
@@ -0,0 +1,160 @@
+)
+
+var (
+	clusterResourcePatchFinalizer = map[string]interface{}{
+		"spec": map[string]interface{}{
+			"cleanupPolicy": map[string]string{
+				"confirmation": "yes-really-destroy-data",
+			},
+		},
+	}
+
+	defaultResourceRemoveFinalizers = map[string]interface{}{
+		"metadata": map[string]interface{}{
+			"finalizers": nil,
+		},
+	}
+)
+
+func DeleteCustomResources(ctx context.Context, clientsets k8sutil.ClientsetsInterface, clusterNamespace string) {
+	err := deleteCustomResources(ctx, clientsets, clusterNamespace)
+	if err != nil {
+		logging.Fatal(err)
+	}
+	logging.Info("done")
+}
+
+func deleteCustomResources(ctx context.Context, clientsets k8sutil.ClientsetsInterface, clusterNamespace string) error {
+	for _, resource := range cephResources {
+		logging.Info(fmt.Sprintf("Getting resources kind %s", resource))
+		items, err := clientsets.ListResourcesDynamically(ctx, cephRookIoGroup, cephRookResourcesVersion, resource, clusterNamespace)
+		if err != nil {
+			if k8sErrors.IsNotFound(err) {
+				logging.Info("the server could not find the requested resource: %s", resource)
+				continue
+			}
+			return err
+		}
+
+		if len(items) == 0 {
+			logging.Info("resource %s was not found on the cluster", resource)
+			continue
+		}
+
+		for _, item := range items {
+			logging.Info(fmt.Sprintf("removing resource %s: %s", resource, item.GetName()))
+			err = clientsets.DeleteResourcesDynamically(ctx, cephRookIoGroup, cephRookResourcesVersion, resource, clusterNamespace, item.GetName())
+			if err != nil {
+				if k8sErrors.IsNotFound(err) {
+					logging.Info(err.Error())
+					continue
+				}
+				return err
+			}
+
+			itemResource, err := clientsets.GetResourcesDynamically(ctx, cephRookIoGroup, cephRookResourcesVersion, resource, item.GetName(), clusterNamespace)
+			if err != nil {
+				if k8sErrors.IsNotFound(err) {
+					logging.Info(err.Error())
+					continue
+				}
+				return err
+			}
+
+			if itemResource != nil {
+				logging.Info(fmt.Sprintf("resource %q is not yet deleted, applying patch to remove finalizer...", itemResource.GetName()))
+				err = updatingFinalizers(ctx, clientsets, itemResource, resource, clusterNamespace)
+				if err != nil {
+					if k8sErrors.IsNotFound(err) {
+						logging.Info(err.Error())
+						continue
+					}
+					return err
+				}
+
+				err = clientsets.DeleteResourcesDynamically(ctx, cephRookIoGroup, cephRookResourcesVersion, resource, clusterNamespace, item.GetName())
+				if err != nil {
+					if k8sErrors.IsNotFound(err) {
+						logging.Info(err.Error())
+						continue
+					}
+					return err
+				}
+			}
+			logging.Info(fmt.Sprintf("resource %s was deleted", item.GetName()))
+		}
+	}
+	return nil
+}
+
+func updatingFinalizers(ctx context.Context, clientsets k8sutil.ClientsetsInterface, itemResource *unstructured.Unstructured, resource, clusterNamespace string) error {
+	patch := defaultResourceRemoveFinalizers
+
+	if itemResource.GetKind() == CephResourceKindCephCluster {
```

Comment:
Will we ever remove the finalizer for the CephCluster CR? If we're force deleting the cluster, at some point we should timeout with the cleanup policy and force delete the CephCluster CR too.

Reply:
As discussed, let's not add a timeout. For now, we will just wait for the cleanup policy to complete the removal.